system-helper/flatpak-system-helper.c

```diff
@@ -190,6 +434,345 @@ get_sender_pid (GDBusMethodInvocation *invocation)
+            {
+              glnx_throw_errno_prefix (&error, "Can't create directory for bindfs mount at %s", user_dir);
+              g_dbus_method_invocation_return_gerror (invocation, error);
+              return TRUE;
+            }
+
+          root_dir_file = g_file_new_for_path (root_dir);
+          child_repo = flatpak_dir_create_child_repo (system, root_dir_file, &child_repo_lock, NULL, &error);
+          if (child_repo == NULL)
+            {
+              g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR, G_DBUS_ERROR_FAILED,
+                                                     "Failed to create OSTree child repo at %s", root_dir);
+              return TRUE;
+            }
+          child_repo_path = g_file_get_path (ostree_repo_get_path (child_repo));
+        }
+
+      g_assert (child_repo_path != NULL);
+
+      new_pull = system_helper_ongoing_pull_new (object, invocation, ref, commit,
+                                                 uid, user_dir, root_dir, child_repo_path);
+
+      /* In bindfs, --force-user forces file-ownership to be seen in the mounted mirrored directory(i.e. user_dir),
+         while --create-for-user enforces the actual owner of the file during file creation on disk.  */
+      arg_user = g_strdup_printf ("--force-user=%"G_GUINT32_FORMAT, uid);
+      new_pull->bindfs = g_subprocess_new (G_SUBPROCESS_FLAGS_NONE,
+                                           &error,
+                                           bindfs_path,
+                                           arg_user,"-f",
+                                           "-o", "nonempty",
+                                           "--create-for-user=root",
+                                           "--create-for-group=root",
+                                           root_dir, user_dir, NULL);
+      if (new_pull->bindfs == NULL)
+        {
+          system_helper_ongoing_pull_teardown (new_pull);
+          g_dbus_method_invocation_return_gerror (invocation, error);
+          return TRUE;
+        }
+
+      G_LOCK (ongoing_pulls);
+      g_ptr_array_add (ongoing_pulls, new_pull);
+      G_UNLOCK (ongoing_pulls);
+
+      g_subprocess_wait_check_async (new_pull->bindfs,
+                                     new_pull->cancellable,
+                                     subprocess_wait_check_cb,
+                                     system_helper_ongoing_pull_ref (new_pull));
+    }
+
+  return TRUE;
+}
+
+static FlatpakSystemHelperOngoingPull *
+get_ongoing_pull_by_dir (const gchar *arg_repo_path)
+{
+  FlatpakSystemHelperOngoingPull *pull = NULL;
+  g_autofree gchar *user_dir = g_path_get_dirname (arg_repo_path);
+
+  G_LOCK (ongoing_pulls);
+  for (guint i = 0; i < ongoing_pulls->len; i++)
+    {
+      pull = g_ptr_array_index (ongoing_pulls, i);
+      if (g_strcmp0 (pull->user_dir_name, user_dir) == 0)
+        break;
+    }
+  G_UNLOCK (ongoing_pulls);
+
+  return pull;
+}
+
```

Comment:
the list of ongoing pulls is not holding a ref on the pull object (otherwise, you couldn't remove the pull from the list as side-effect of finalizing the object. it seems to me that to be threadsafe, this function needs to return a ref, taken under the lock.

Reply:
i have changed the approach here. We basically remove the ongoing pull from the array and steal the reference, return it to the respective caller which takes care of the freeing it. Does it makes sense @matthiasclasen ? More details in this [comment/thread](https://github.com/flatpak/flatpak/pull/2342#discussion_r237066874)?